pkg/controller/lambda/function/setup.go

```diff
@@ -0,0 +1,516 @@
+	}
+
+	// Connection settings for an Amazon EFS file system.
+	if !isUpToDateFileSystemConfigs(cr, obj) {
+		return false, nil
+	}
+
+	if aws.StringValue(cr.Spec.ForProvider.Handler) != aws.StringValue(obj.Configuration.Handler) {
+		return false, nil
+	}
+
+	if aws.StringValue(cr.Spec.ForProvider.KMSKeyARN) != aws.StringValue(obj.Configuration.KMSKeyArn) {
+		return false, nil
+	}
+
+	// The function's layers (https://docs.aws.amazon.com/lambda/latest/dg/configuration-layers.html).
+	// The generator is creating layers with type of
+	// 	Layers []*string `json:"layers,omitempty"`
+	// Instead of []*svcsdk.Layer
+	// We can't properly implement a comparison until that is fixed
+
+	// set default
+	if aws.Int64Value(cr.Spec.ForProvider.MemorySize) != aws.Int64Value(obj.Configuration.MemorySize) {
+		return false, nil
+	}
+
+	if aws.StringValue(cr.Spec.ForProvider.Role) != aws.StringValue(obj.Configuration.Role) {
+		return false, nil
+	}
+
+	if aws.StringValue(cr.Spec.ForProvider.Runtime) != aws.StringValue(obj.Configuration.Runtime) {
+		return false, nil
+	}
+
+	if aws.Int64Value(cr.Spec.ForProvider.Timeout) != aws.Int64Value(obj.Configuration.Timeout) {
+		return false, nil
+	}
+
+	if aws.StringValue(cr.Spec.ForProvider.TracingConfig.Mode) != aws.StringValue(obj.Configuration.TracingConfig.Mode) {
+		return false, nil
+	}
+
+	if !isUpToDateSecurityGroupIDs(cr, obj) {
+		return false, nil
+	}
+
+	addTags, removeTags := diffTags(cr.Spec.ForProvider.Tags, obj.Tags)
+	return len(addTags) == 0 && len(removeTags) == 0, nil
+
+}
+
+// isUpToDateEnvironment checks if FunctionConfiguration EnvironmentResponse Variables are up to date
+func isUpToDateEnvironment(cr *svcapitypes.Function, obj *svcsdk.GetFunctionOutput) bool {
+	// Handle nil pointer refs
+	envVars := map[string]*string{}
+	awsVars := map[string]*string{}
+	if cr.Spec.ForProvider.Environment != nil &&
+		cr.Spec.ForProvider.Environment.Variables != nil {
+		envVars = cr.Spec.ForProvider.Environment.Variables
+	}
+	if obj.Configuration.Environment != nil &&
+		obj.Configuration.Environment.Variables != nil {
+		awsVars = obj.Configuration.Environment.Variables
+	}
+
+	return compareMapStringStringPointer(envVars, awsVars)
+}
+
+// compareSliceStringPointer checks if 2 []*string contain the same values
+func compareSliceStringStringPointer(cr []*string, obj []*string) bool {
+	if len(cr) != len(obj) {
+		return false
+	}
+	for i, v := range cr {
+		if aws.StringValue(v) != aws.StringValue(obj[i]) {
+			return false
+		}
+	}
+	return true
+}
+
+// compareMapStringStringPointer checks if 2 map[string]*string contain the same keys and values
+func compareMapStringStringPointer(cr map[string]*string, obj map[string]*string) bool {
+	for k, v := range cr {
+		if aws.StringValue(obj[k]) == aws.StringValue(v) {
+			continue
+		}
+		return false
+	}
+	for k, v := range obj {
+		if aws.StringValue(cr[k]) == aws.StringValue(v) {
+			continue
+		}
+		return false
+	}
+	return true
+}
```

Comment:
```suggestion
```

nitpick, you don't have to do it: What do you think about having loops in `isUpToDateEnvironment` and `isUpToDateSecurityGroupIDs` that fills up `map[string]string`s and then compare two maps using `cmp.Equal`? That's take care of pointers as well as ordering issue with slices where the elements are same but order is different, hence false negative.

Reply:
Or an approach similar to `isUpToDateFileSystemConfigs`. Dealing with pointers in more than one place sort of makes me feel unsafe because of the panic possibility.

Reply:
good point - i've updated based on the comment to use `cmp.Equal` which works great.  IMO now that I see  this, for all comparisons should always use `cmp.Equal` unless there is a compelling reason not to.